### PR TITLE
Support rechunking to a frequency.

### LINKED ISSUE #7559: Support specifying chunk sizes using labels (e.g. frequency string)
### Is your feature request related to a problem?

`dask.dataframe` [supports](https://docs.dask.org/en/stable/generated/dask.dataframe.DataFrame.repartition.html) repartitioning or rechunking using a frequency string (`freq` kwarg).

I think this would be a useful addition to `.chunk`. It would help with some groupby problems ([as suggested in this comment](https://github.com/pydata/xarray/issues/2237#issuecomment-398581618)) and generally make a few problems amenable to blockwise/map_blocks solutions.

### Describe the solution you'd like

1. One solution is to allow `.chunk(lon=5, time="MS")`. There is some ugliness in that this syntax mixes up integer index values (`lon=5`) and a label-based frequency string `time="MS"`
2. So perhaps a second method `chunk_by_labels` would be useful where `chunk_by_labels(lon=5, time="MS")` would rechunk the data so that a single chunk contains 5° of longitude points and a month of time. Alternative this could be `.chunk(lon=5, time="MS", by="labels")`

### Describe alternatives you've considered

Have the user do this manually but that's kind of annoying, and a bit advanced.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #7559
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

Now uses syntax `ds.chunk(time=TimeResampler("YE"))`

### CODE REVIEW

**Comment by shoyer** [comment]
Are frequency strings unambiguous? Rechunking already supports memory sizes for Dask using strings.

**Comment by dcherian** [comment]
Ready for review!

The syntax is `ds.chunk(time=TimeResampler("YE"))`.

**Comment by dcherian** [comment]
This is pretty cool, and could use a review!

File: doc/user-guide/dask.rst [hunk 1]
```diff
@@ -296,6 +296,12 @@ loaded into Dask or not:
 Automatic parallelization with ``apply_ufunc`` and ``map_blocks``
 -----------------------------------------------------------------
 
+.. tip::
+
+   Some problems can become embarassingly parallel and thus easy to parallelize
+   automatically by rechunking to a frequency, e.g. ``ds.chunk(time=TimeResampler("YE"))``.
```
Comment (spencerkclark) [inline doc/user-guide/dask.rst:302]: I think currently the only way to import / use `TimeResampler` from the public API is via either:
```
>>> import xarray
>>> xarray.groupers.TimeResampler
```
or
```
>>> from xarray import groupers
>>> groupers.TimeResampler
```
In other words it does not seem possible to import the name `TimeResampler` alone unless one does the following (which I think we would discourage):
```
>>> from xarray.core.groupers import TimeResampler
```
We should probably do something to address this or update the documentation.

**Review by spencerkclark — APPROVED** [review]
This looks great @dcherian!  My only suggestion is to make the documentation a little clearer about how to import this new object.

File: doc/user-guide/dask.rst (hunk 1 above)
Comment (dcherian) [reply doc/user-guide/dask.rst:302]: Nice catch, how would we enable `from xarray.groupres import TimeResampler`? Moving the file up from `core/`?

Comment (dcherian) [reply doc/user-guide/dask.rst:302]: Fixed by #9289

Comment (spencerkclark) [reply doc/user-guide/dask.rst:302]: Great, thanks!

**Pull request merged**